feat: Input 컴포넌트 (props를 통해 구현)

newminkyung opened the pull request:
### 이슈 번호

depromeet/13th-4team-client#11

### 작업 분류

- [ ] 버그 수정
- [x] 신규 기능
- [ ] 프로젝트 구조 변경

### 작업 상세 내용

- `TextField`, `TextAreaField` 구현 (네이밍은 피그마 참고)
- `TextLengthMessage`: 사용자가 입력한 글자 수 / 최대 입력 가능 글자 수` 를 보여주는 컴포넌트 구현
- `ErrorMessage`: 에러메시지를 보여주는 컴포넌트 구현

#### 1. 스타일링 방식: css 파일에 class명 정의
> 더 좋은 의견 있으시면 리뷰로 남겨주세요!
  - global.css에 `form`, `form-typed`, `form-error`로 정의되어있습니다
  - form-typed, form-error는 form 스타일에 의존합니다.
  - TextField, TextareaField, SelectField, PickerField 모두 동일한 스타일링 방식을 사용하고 있어서 global.css에 정의했습니다
 

#### 2. Input 컴포넌트 구현 방식
- props로 Input 컴포넌트틀 결정합니다.
- 장점: 사용부에서 props를 전달해서 쉽게 사용 가능
- 단점: 컴포넌트의 확장성, 유연성이 떨어짐 (mode가 추가될 때마다 컴포넌트에 추가해줘야함)

### 사용 방법

#### TextAreaField

<img width="300" alt="image" src="https://github.com/depromeet/13th-4team-client/assets/80238096/1bec9062-be4d-4593-a4f2-c3b31a45f627">


``` tsx
<TextAreaField placeholder="ex.개발자와 협업 역량을 쌓기 위해 it 동아리에 들어감" />
<TextAreaField showCount maxLength={100} placeholder="ex.개발자와 협업 역량을 쌓기 위해 it 동아리에 들어감" />
<TextAreaField disabled placeholder="ex.개발자와 협업 역량을 쌓기 위해 it 동아리에 들어감" />
<TextAreaField error errorMessage="필수 입력사항입니다" />
```

- Props

| props | 값 | 설명 |
|--------|--------|--------|
| placeholder | string | placeholder  |
| showCount / maxLength | boolean / number | `TextLengthMessage` 컴포넌트를 보여주는 값입니다. <br /> showCount 사용 시 , maxLength도 함께 전달해야합니다.|
|disabled | boolean | disabled |
| error / errorMessage | boolean / string | `ErrorMessage` 컴포넌트를 보여주는 값입니다. <br /> error 사용 시, errorMessage도 함께 전달해야합니다. |

#### TextField

<img width="300" alt="image" src="https://github.com/depromeet/13th-4team-client/assets/80238096/b8a9d628-8b2f-4256-8bee-e2b7eb55611c">


``` tsx
<TextField placeholder="abc@gmail.com" />
<TextField disabled placeholder="abc@gmail.com" />
<TextField error errorMessage="필수 입력사항입니다" />
<TextField mode="naming" maxLength={10} placeholder="abc@gmail.com" />
<TextField mode="password" />
<TextField mode="email" emailDomain="@gmail.com" />
<TextField mode="chip" placeholder="abc@gmail.com" />
```

- Props
> TextField와 동일한 props는 생략

| props | 값 | 설명 |
|--------|--------|--------|
| mode | default <br /> naming <br /> password <br /> email <br /> chip |  첨부 사진 참고 부탁드립니다.   |
| emailDomain | string | `mode`가 `email`인 경우, emailDomain을 함께 전달해야합니다. |

newminkyung committed:
feat: ErrorMessage 컴포넌트

newminkyung committed:
feat: TextLengthMessage 컴포넌트

newminkyung committed:
feat: TextField 컴포넌트

newminkyung committed:
feat: TextAreaField 컴포넌트

newminkyung committed:
feat: demo 페이지

sangbooom commented on src/components/Input/TextField/TextField.tsx:
```diff
@@ -0,0 +1,57 @@
+  return (
+    <div className="relative">
+      <label htmlFor="">{mode === 'chip' && <Chip />}</label>
+      <input
+        type={mode === 'password' ? 'password' : 'text'}
+        className={`form w-full ${isTyped ? 'form-typed' : ''} ${error ? 'form-error' : ''}`}
+        {...props}
+      />
```
error 상황일경우 input에 hover했을때와 입력중일때는 보더라인이 검정색인데 에러메시지는 떠있어서 개인적으로 ui가 조금 어색해 보입니다! 입력한 순간 에러메시지를 없애도 좋고 아니면 input에 hover했을때와 입력중일때는 계속 보더라인을 빨간색라인으로 유지해도 좋을 것 같다는 의견 드립니다~ (디자이너팀과 상의해보아요 🙌)

newminkyung replied:
안그래도 지선님이랑 얘기해보았는데 에러일 때 보더라인은 그대로(hover, focus 스타일) 유지해도 좋을 것 같다고 하셨어요!
에러 메시지를 없애는 방향으로 변경하면 좋을 것 같습니다👍

sangbooom commented on src/components/Input/TextField/TextField.tsx:
```diff
@@ -0,0 +1,57 @@
+import { ErrorMessage } from '../ErrorMessage';
+import { TextLengthMessage } from '../TextLengthMessage';
+
+type TextFieldProps = ComponentPropsWithRef<'input'> & {
+  mode?: 'default' | 'naming' | 'password' | 'email' | 'chip';
+  size?: 'small' | 'middle';
```
size 속성은 input태그에 이미 있는 속성이네요! 다른 이름으로 정의하는게 좋을 것 같아요!

> https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/size

sangbooom commented on src/app/demo/input/page.tsx:
```diff
@@ -0,0 +1,27 @@
+        <TextField disabled placeholder="abc@gmail.com" />
+        <TextField isTyped placeholder="abc@gmail.com" />
+        <TextField error errorMessage="필수 입력사항입니다" />
+        <TextField mode="naming" maxLength={10} placeholder="abc@gmail.com" />
```
아니 이 텍스트필드 왜 maxLength 안먹죠,,? 🧐 string으로 넘기면 되는데 number로 넘기면 안되는 이유를 아시나요 😂

newminkyung replied:
오잉..? maxLength 타입이 `number | undefined` 인데 오히려 string이 되나요..?!

sangbooom replied:
지금 코드로 maxLength가 정상 작동안해서 이것저것 해보다가 `maxLength={10}` 을 `maxLength="10"`으로 변경하니까 되더라구요..? 근데 이 방법보단 그냥 useInput 훅 하나 만들고 changeHandler 에서 판별하는게 더 깔끔하게 처리하는게 좋아보이네요
```ts
// hooks/useInput.ts
import { useState, useCallback, ChangeEvent } from 'react';

// eslint-disable-next-line unused-imports/no-unused-vars
type onChangeType = (event: ChangeEvent<HTMLTextAreaElement | HTMLInputElement>) => void;

const useInput = (initialValue = '') => {
  const [value, setValue] = useState(initialValue);
  const handler = useCallback(({ target }: ChangeEvent<HTMLTextAreaElement | HTMLInputElement>) => {
    const max = Number(target.getAttribute('maxLength'));

    if (max && target.value.length > max) {
      target.value = target.value.slice(0, max);
    }

    setValue(target.value);
  }, []);
  return [value, handler, setValue] as [string, onChangeType, typeof setValue];
};

export default useInput;


// 사용부 

const [text, onChangeText] = useInput('');
 ....
        <TextField
          maxLength={10}
          value={text}
          onChange={onChangeText}
          ....
        />
```

전 그냥 다른 방법으로 해결한거라서 근본적인 해결 방법이 있다면 공유부탁드립니다~~

newminkyung replied:
하하.. maxLength 디스트럭처링하고 나서 input attribute로 안넣어줬읍니다.. 🗿

근데 `maxLength="10"`으로 적용했는데 에러 발생해서 .. 이 부분은 왜 작동하는지 잘 모르겠어요..

sangbooom replied:
> 하하.. maxLength 디스트럭처링하고 나서 input attribute로 안넣어줬읍니다.. 🗿

@minkyung00 넵 그부분을 해결하고나서 이후의 상황을 말씀드린겁니다! 넣고나서는 잘 작동하시나요??

newminkyung replied:
@sangbooom 아 네네! 근데 저는 `maxLength="10"` 해봤는데 안되더라구요..

sangbooom replied:
지금은 잘 작동하네요~

sangbooom commented on src/components/Input/TextField/TextField.tsx:
```diff
@@ -0,0 +1,57 @@
+type TextFieldProps = ComponentPropsWithRef<'input'> & {
+  mode?: 'default' | 'naming' | 'password' | 'email' | 'chip';
+  size?: 'small' | 'middle';
+  isTyped?: boolean;
+  error?: boolean;
+  errorMessage?: string;
+  emailDomain?: string;
```
jsdoc으로 설명 추가해주시면 감사하겠습니다! isTyped나 emailDomain은 어떻게 사용하는지 구현부를 자세히 봐야될것같아서요!

newminkyung committed:
chore: TextField jsdoc 추가

sangbooom reviewed (commented):
전 개인적으로 현재 디자인가이드의 input은 props를 통한 구현이 더 적절해보입니다! 
이유는 지금 input 스펙으로는 굳이 유연성과 확장성에 집중하는것보단 props를 통해 쉽게 컨트롤 할 수 있도록 잘 만든다면 사용성 측면에선 우위일 것 같아서요! 다른분들의 의견도 궁금하네요~

newminkyung committed:
feat: isTyped props 삭제

- isTyped: value가 있으면 true인 속성
- value props로 판단할 수 있는 속성이기 때문에 삭제

newminkyung committed:
fix: attribute에 maxLength 추가

newminkyung committed:
feat: inputSize props 삭제

- 불필요한 속성이여서 삭제

newminkyung committed:
chore: TextAreaField jsdoc 추가

newminkyung committed:
fix: demo 페이지에서 삭제된 props 제거

newminkyung committed:
style: TexAreaField 스타일링 수정

newminkyung committed:
chore: storybook tailwind 관련 설정

newminkyung committed:
feat: tailwind-merge 적용

newminkyung committed:
feat: ErrorIcon 추가

newminkyung committed:
Merge branch 'main' of https://github.com/depromeet/13th-4team-client into feat/Input

newminkyung committed:
feat: TextAreaField 스토리 작성

newminkyung committed:
fix: storybook alias 설정 오류 해결

- baseUrl 추가
- 참고: https://github.com/storybookjs/storybook/issues/21901#issuecomment-1538802687

newminkyung committed:
chore: storybook tailwindcss 설정 삭제

- chakra ui 스타일 적용이 안되는 현상이 발생하여 관련 설정을 삭제

newminkyung committed:
feat: Icon 컴포넌트 분리

newminkyung committed:
style: TextLengthMessage 컴포넌트 스타일링 수정

- 글자 색, 글자 크기 외에 별도의 스타일링은 사용부에서 추가
- absolute시 height가 안잡혀서 float-right로 변경

stae1102 commented on src/app/demo/input/page.tsx:
```diff
@@ -0,0 +1,28 @@
+  return (
+    <>
+      <div className="flex flex-col gap-4 p-10">
+        <TextAreaField placeholder="ex.개발자와 협업 역량을 쌓기 위해 it 동아리에 들어감" />
```
백엔드에서는 const 폴더 같은 거 만들어서 동일한 string들은 하나의 상수로 처리했습니다. 이게 다른 곳에서도 많이 쓰일 수도 있고, human error도 발생할 수 있어서 다른 파일에 상수로 값을 평가하고 export 해보는 것도 추천드려유

newminkyung replied:
지존 꼼꼼👍

stae1102 reviewed (commented):
지존레전드

stae1102 reviewed (commented):
지존레전드

stae1102 reviewed (commented):
지존레전드
